src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/TransactionValidator.cs

```diff
@@ -0,0 +1,45 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using static Hl7.Fhir.Model.Bundle;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources.Bundle
+{
+    public static class TransactionValidator
+    {
+        public static void ValidateTransaction(HashSet<string> resourceIdList, EntryComponent entry)
+        {
+            if (entry.Request.Method != HTTPVerb.GET)
+            {
+                string resourceId = GetResourceUrl(entry);
+
+                if (resourceIdList.Contains(resourceId))
+                {
+                    throw new RequestNotValidException(string.Format(Api.Resources.ResourcesMustBeUnique, resourceId));
+                }
+
+                resourceIdList.Add(resourceId);
+            }
+        }
+
+        private static string GetResourceUrl(EntryComponent component)
```

Comment:
There's also some actions that aren't GETs that realistically we don't have to worry about. Operations via `POST` is an example. https://www.hl7.org/fhir/http.html#summary